lib/SILOptimizer/Utils/ShrinkBorrowScope.cpp

```diff
@@ -24,295 +33,371 @@ using namespace swift;
-      [&deadBlocks](SILBasicBlock *block) -> bool {
-    return llvm::all_of(block->getSuccessorBlocks(), [=](auto *successor) {
-      return deadBlocks.contains(successor);
-    });
-  };
+private:
+  friend class BackwardReachability<DataFlow>;
 
-  while (!worklist.empty()) {
-    auto *block = worklist.pop_back_val();
-    auto *startingInstruction = startingInstructions.lookup(block);
-    if (!startingInstruction && !hasOnlyDeadSuccessors(block)) {
+  bool hasReachableBegin(SILBasicBlock *block) {
+    return result.hoistingReachesBeginBlocks.contains(block);
+  }
+
+  void markReachableBegin(SILBasicBlock *block) {
+    result.hoistingReachesBeginBlocks.insert(block);
+  }
+
+  void markReachableEnd(SILBasicBlock *block) {
+    result.hoistingReachesEndBlocks.insert(block);
+  }
+
+  bool checkReachableBarrier(SILInstruction *inst);
+};
+
+/// Whether the specified value is %lifetime or its iterated copy_value.
+///
+/// In other words, it has to be a simple extended def of %lifetime.
+bool isSimpleExtendedIntroducerDef(Context const &context, SILValue value) {
+  while (true) {
+    auto *instruction = value.getDefiningInstruction();
+    if (!instruction)
+      return false;
+    if (instruction == &context.introducer)
+      return true;
+    if (auto *cvi = dyn_cast<CopyValueInst>(instruction)) {
+      value = cvi->getOperand();
       continue;
     }
-    for (auto *successor : block->getSuccessorBlocks()) {
-      barredBlocks.erase(successor);
-    }
+    return false;
+  }
+}
 
-    // We either have processed all successors of block or else it is a block
-    // which contained one of the original scope-ending instructions.  Scan the
-    // block backwards, looking for the first deinit barrier.  If we've visited
-    // all successors, start scanning from the terminator.  If the block
-    // contained an original scope-ending instruction, start scanning from it.
-    SILInstruction *instruction =
-        startingInstruction ? startingInstruction : block->getTerminator();
-    if (!startingInstruction) {
-      // That there's no starting instruction means that this this block did not
-      // contain an original introducer.  It was added to the worklist later.
-      // At that time, it was checked that this block (along with all that
-      // successor's other predecessors) had a terminator over which the borrow
-      // scope could be shrunk.  Shrink it now.
-      bool hoisted = tryHoistOverInstruction(block->getTerminator());
-      assert(hoisted);
-      (void)hoisted;
-    }
-    SILInstruction *barrier = nullptr;
-    while ((instruction = instruction->getPreviousInstruction())) {
-      if (!tryHoistOverInstruction(instruction)) {
-        barrier = instruction;
-        break;
-      }
+bool DataFlow::checkReachableBarrier(SILInstruction *instruction) {
+  if (instruction == &context.introducer) {
+    result.barriers.push_back(instruction);
+    return true;
+  }
+  if (uses.containsEnd(instruction)) {
```

Comment:
I haven't figure out yet why you need the `containsEnd` method. If reachability starts backward from the scope ending instructions, then it should never need to check for them.

Reply:
It was just implemented that way for simplicity and convenience at the time. I suggest changing reachability so it doesn't check the last use for a barrier. The client can always do that explicitly if needed. Then `initLastUse` should have a comment that reachability assumes it is a non-barrier, because that is pretty subtle.

Alternatively, just add the end_borrows to the users set.

`checkReachableBarrier` is the critical path for the data flow. It should only do constant-time checks and access as few data structures as possible.

Reply:
> checkReachableBarrier is the critical path for the data flow. It should only do constant-time checks and access as few data structures as possible.

Makes sense!  I'll see about tweaking reachability like that.

Reply:
Made that change along with some others to deal with barrier terminators in https://github.com/apple/swift/pull/41358/commits/05e3c5a8e84e87886652f5f9d1c6e0c7276a7ddd .